### ISSUE TITLE
False positive for lambda within a function

### ISSUE THREAD

**Issue by michaelbrewer** [python/mypy#10426]
**Related Bug**

Could be a duplicate of:
- https://github.com/python/mypy/issues/4297

**Bug Report**

When running mypy with a function returning an Optional the contains of a lambda expression is interpreted as optional and a type error is raised

**To Reproduce**

Run `mypy` against the following file:

```python3
from typing import Optional

animals = ["cow"]


def find1(name: str) -> Optional[str]:
    return next(
        filter(
            lambda animal: animal.upper() == name,
            animals,
        ),
        None,
    )
```

**Expected Behavior**

No type errors should be returned

**Actual Behavior**

```shell
$ mypy bug.py
bug.py:9: error: Item "None" of "Optional[str]" has no attribute "upper"
```

NOTE: Same error with returning "Union[str, None]"
NOTE: Removing the Optional[str] stops the error, but that defaults the point 

```python3
def find1(name: str):
    return next(
        filter(
            lambda animal: animal.upper() == name,
            animals,
        ),
        None,
    )
```

**Your Environment**

- Mypy version used: 0.812
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9.4
- Operating system and version: MacOS 10.15.7

**Comment by hauntsaninja** [comment]
From the title I thought it was going to be a dupe of https://github.com/python/mypy/issues/2608, but it's not!

The dumb workaround is:
```
from typing import Optional

animals = ["cow"]


def find1(name: str) -> Optional[str]:
    ret = next(
        filter(
            lambda animal: animal.upper() == name,
            animals,
        ),
        None,
    )
    return ret
```

I think this boils down to the fact that mypy does complicated things to infer types for lambdas. If there's an annotation, it takes into account what type the expression is expected to be to try and understand the lambda. E.g., if you were to change `ret = next(...)` to `ret: Optional[str] = next(...)` you'd also get an error.

**Comment by michaelbrewer** [comment]
Thanks @hauntsaninja , my lint tools would have suggested to immediately return :). So i guess it is a coin flip.
